File: core/src/main/scala/com/graphbrain/eco/nodes/patterns/VarPatternElem.scala
package com.graphbrain.eco.nodes.patterns

class VarPatternElem(val name: String,
                     val possiblePOS: Array[String]=Array[String](),
                     val necessaryPOS: Array[String]=Array[String](),
                     val forbiddenPOS: Array[String]=Array[String]())
  extends PatternElem {

  var intervals = List[(Int, Int)]()
  var curInterval = -1
  var intEnd = -1

  override protected def priority: Int = possiblePOS.length + necessaryPOS.length + forbiddenPOS.length

  private def possibleWord(pos: Int): Boolean = {
    if (possiblePOS.length == 0)
      return true

    val wordPOS = sentence.words(pos).pos
    val b = possiblePOS.exists(p => wordPOS.startsWith(p))
    //println(pos)
    //println(sentence.words(pos))
    //println(possiblePOS.reduceLeft(_ + ", " + _))
    //println(b)
    b
  }

  override protected def onPreProcess() = {
    intervals = List[(Int, Int)]()

    var intStart = startMin
    var intEnd = -1

    while (intStart <= startMax) {
      if (possibleWord(intStart)) {
        intEnd = intStart
        while ((intEnd <= endMax) && possibleWord(intEnd))
          intEnd += 1

        intervals = (intStart, intEnd - 1) :: intervals

        intStart = intEnd + 1
      }
      else {
        intStart += 1
      }
    }

    intervals = intervals.reverse
    //println(endMax)
    //println(intervals)
  }

  override protected def onSetSentence() = {
    if (prevElem == null) {
      startMin = 0
      startMax = 0
    }
    else {
      startMin = prevElem.endMin + 1
      startMax = prevElem.endMax + 1
    }

    val slen = sentence.length
    val remaining = elemCount - elemPos - 1

    endMin = if (remaining == 0) slen - 1 else startMin
    endMax = slen - 1 - remaining
  }

  private def curIntervalValid: Boolean = {
    if ((necessaryPOS.length == 0) && (forbiddenPOS.length == 0))
      return true

    if (necessaryPOS.length > 0) {
      val necessary = necessaryPOS.forall(np =>
        (start to end).exists(i =>
          sentence.words(i).pos.startsWith(np)))

      if (!necessary) return false
    }

    if (forbiddenPOS.length > 0)
      forbiddenPOS.forall(np =>
        !(start to end).exists(i =>
          sentence.words(i).pos.startsWith(np)))
    else
      true
  }

  override def onNext(): Boolean = {
    var found = false

    while(!found) {
      if (!step)
        return false

      if (curIntervalValid)
        found = true
    }

    //println(this)

    true
  }

  private def setCurInterval(n: Int): Boolean = {
    if (n >= intervals.length)
      return false

    curInterval = n
    intEnd = Math.min(curEndMax, intervals(curInterval)._2)

    start = Math.max(curStartMin, intervals(curInterval)._1)
    if (start > curStartMax)
      return false

    end = Math.max(curEndMin, start)

    if (end > intEnd)
      return false

    true
  }

  private def step: Boolean = if (start < 0) {
    var firstInterval = 0

    // look for first interval that can fit
    while ((firstInterval < intervals.length)
      && (intervals(firstInterval)._2 < curStartMin))
      firstInterval += 1

    setCurInterval(firstInterval)
  }
  else {
    end += 1
    if (end > intEnd) {
      start += 1
      if ((start > intervals(curInterval)._2) || (start > curStartMax)) {
        if (!setCurInterval(curInterval + 1))
          return false
      }
      else {
        Math.max(curEndMin, start)
      }
    }
    true
  }

  def curWords = sentence.slice(start, end)

  override def toString = if (sentence == null)
    name + ":" +
      (if (possiblePOS.isEmpty) "" else possiblePOS.reduceLeft(_ + "|" + _)) +
      "+" +
      (if (necessaryPOS.isEmpty) "" else necessaryPOS.reduceLeft(_ + "|" + _)) +
      "-" +
      (if (forbiddenPOS.isEmpty) "" else forbiddenPOS.reduceLeft(_ + "|" + _))
  else
    name + " = '" + sentence.slice(start, end) + "'"
}

File: core/src/main/java/com/graphbrain/eco/nodes/patterns/VarPatternElem.java
package com.graphbrain.eco.nodes.patterns;

import com.graphbrain.eco.Words;

import java.util.Vector;

public class VarPatternElem extends PatternElem {
    private String name;
    private String[] possiblePOS;
    private String[] necessaryPOS;
    private String[] forbiddenPOS;

    private Vector<Integer[]> intervals;
    private int curInterval;
    private int intEnd;

    public VarPatternElem(String name,
        String[] possiblePOS,
        String[] necessaryPOS,
        String[] forbiddenPOS) {

        super();

        this.name = name;
        this.possiblePOS = possiblePOS;
        this.necessaryPOS = necessaryPOS;
        this.forbiddenPOS = forbiddenPOS;

        intervals = new Vector<Integer[]>();
        curInterval = -1;
        intEnd = -1;
    }

    public VarPatternElem(String name) {
        this(name, new String[]{}, new String[]{}, new String[]{});
    }

    @Override
    protected int priority() {
        return possiblePOS.length + necessaryPOS.length + forbiddenPOS.length;
    }

    private boolean possibleWord(int pos) {
        if (possiblePOS.length == 0)
            return true;

        String wordPOS = sentence.getWords()[pos].getPos();
        boolean b = false;

        for (String p : possiblePOS) {
            if (wordPOS.startsWith(p)) {
                b = true;
                break;
            }
        }

        return b;
    }

    @Override
    protected void onPreProcess() {
        intervals = new Vector<Integer[]>();

        int intStart = startMin;

        while (intStart <= startMax) {
            if (possibleWord(intStart)) {
                int intEnd = intStart;
                while ((intEnd <= endMax) && possibleWord(intEnd))
                    intEnd += 1;

                Integer[] interval = {intStart, intEnd - 1};
                intervals.add(interval);

                intStart = intEnd + 1;
            }
            else {
                intStart += 1;
            }
        }
    }

    @Override
    protected void onSetSentence() {
        if (prevElem == null) {
            startMin = 0;
            startMax = 0;
        }
        else {
            startMin = prevElem.endMin + 1;
            startMax = prevElem.endMax + 1;
        }

        int slen = sentence.length();
        int remaining = elemCount - elemPos - 1;

        if (remaining == 0)
            endMin = slen - 1;
        else
            endMin = startMin;

        endMax = slen - 1 - remaining;
    }

    private boolean curIntervalValid() {
        if ((necessaryPOS.length == 0) && (forbiddenPOS.length == 0))
            return true;

        if (necessaryPOS.length > 0) {
            boolean necessary = true;

            for (String np : necessaryPOS) {
                boolean test = false;
                for (int i = start; i <= end; i++) {
                    if (sentence.getWords()[i].getPos().startsWith(np)) {
                        test = true;
                        break;
                    }
                }

                if (!test) {
                    necessary = false;
                    break;
                }
            }

            if (!necessary)
                return false;
        }

        if (forbiddenPOS.length > 0) {
            for (String fp : forbiddenPOS) {
                for (int i = start; i <= end; i++) {
                    if (sentence.getWords()[i].getPos().startsWith(fp)) {
                        return false;
                    }
                }
            }
        }
        else {
            return true;
        }

        return true;
    }

    @Override
    public boolean onNext() {
        boolean found = false;

        while(!found) {
            if (!step())
                return false;

            if (curIntervalValid())
                found = true;
        }

        return true;
    }

    private boolean setCurInterval(int n) {
        if (n >= intervals.size())
            return false;

        curInterval = n;
        intEnd = Math.min(curEndMax(), intervals.get(curInterval)[1]);

        start = Math.max(curStartMin(), intervals.get(curInterval)[0]);
        if (start > curStartMax())
            return false;

        end = Math.max(curEndMin(), start);

        return end <= intEnd;
    }

    private boolean step() {

        if (start < 0) {
            int firstInterval = 0;

            // look for first interval that can fit
            while ((firstInterval < intervals.size())
                    && (intervals.get(firstInterval)[1] < curStartMin())) {

                firstInterval += 1;
            }

            return setCurInterval(firstInterval);
        }
        else {
            end += 1;
            if (end > intEnd) {
                start += 1;
                if ((start > intervals.get(curInterval)[1]) || (start > curStartMax())) {
                    if (!setCurInterval(curInterval + 1))
                        return false;
                }
                /*
                else {
                    Math.max(curEndMin(), start);
                }*/
            }
            return true;
        }
    }

    public Words curWords() {
        return sentence.slice(start, end);
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder(100);

        sb.append(name);

        if ((possiblePOS.length > 0) || (necessaryPOS.length > 0) || (forbiddenPOS.length > 0)) {
            sb.append(":");

            boolean first = true;
            for (String x : possiblePOS) {
                if (first) {
                    first = false;
                }
                else {
                    sb.append("|");
                }
                sb.append(x);
            }

            first = true;
            for (String x : necessaryPOS) {
                if (first) {
                    first = false;
                }
                else {
                    sb.append("|");
                }
                sb.append("+");
                sb.append(x);
            }

            first = true;
            for (String x : forbiddenPOS) {
                if (first) {
                    first = false;
                }
                else {
                    sb.append("|");
                }
                sb.append("-");
                sb.append(x);
            }
        }

        if (sentence != null) {
            sb.append(" = '");
            sb.append(sentence.slice(start, end));
            sb.append("'");
        }

        return sb.toString();
    }

    public String getName() {
        return name;
    }
}